core/src/main/java/com/linecorp/armeria/common/stream/ConcatPublisherStreamMessage.java

```diff
@@ -0,0 +1,280 @@
+        this.sources = sources;
+    }
+
+    @Override
+    public boolean isOpen() {
+        return sources.isOpen();
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return sources.isEmpty();
+    }
+
+    @Override
+    public long demand() {
+        final OuterSubscriber<T> outerSubscriber = this.outerSubscriber;
+        if (outerSubscriber == null) {
+            return 0;
+        }
+
+        final StreamMessage<? extends T> currentPublisher = outerSubscriber.currentPublisher;
+        if (currentPublisher == null) {
+            return 0;
+        }
+
+        return currentPublisher.demand();
+    }
+
+    @Override
+    public CompletableFuture<Void> whenComplete() {
+        return sources.whenComplete();
+    }
+
+    @Override
+    public void subscribe(Subscriber<? super T> subscriber, EventExecutor executor) {
+        subscribe(subscriber, executor, EMPTY_OPTIONS);
+    }
+
+    @Override
+    public void subscribe(Subscriber<? super T> subscriber, EventExecutor executor,
+                          SubscriptionOption... options) {
+        requireNonNull(subscriber, "subscriber");
+        requireNonNull(executor, "executor");
+        requireNonNull(options, "options");
+
+        final InnerSubscriber<T> innerSubscriber = new InnerSubscriber<>(subscriber, options);
+        final OuterSubscriber<T> outerSubscriber = new OuterSubscriber<>(innerSubscriber, executor);
+        this.outerSubscriber = outerSubscriber;
+        subscriber.onSubscribe(innerSubscriber);
+        sources.subscribe(outerSubscriber, executor, options);
+    }
+
+    @Override
+    public void abort() {
+        sources.abort();
+    }
+
+    @Override
+    public void abort(Throwable cause) {
+        sources.abort(cause);
+    }
+
+    private static final class OuterSubscriber<T> implements Subscriber<Publisher<? extends T>> {
+
+        @SuppressWarnings("rawtypes")
+        private static final AtomicReferenceFieldUpdater<OuterSubscriber, Subscription> upstreamUpdater =
+                AtomicReferenceFieldUpdater.newUpdater(OuterSubscriber.class, Subscription.class, "upstream");
+
+        private final InnerSubscriber<T> innerSubscriber;
+        private final EventExecutor executor;
+
+        private boolean completed;
+        private boolean inSubscribe;
+
+        @Nullable
+        private volatile StreamMessage<? extends T> currentPublisher;
+        @Nullable
+        private volatile Subscription upstream;
+
+        OuterSubscriber(InnerSubscriber<T> innerSubscriber, EventExecutor executor) {
+            this.innerSubscriber = innerSubscriber;
+            this.executor = executor;
+            innerSubscriber.setOuterSubscriber(this);
+        }
+
+        @Override
+        public void onSubscribe(Subscription subscription) {
+            requireNonNull(subscription, "subscription");
+            if (upstreamUpdater.compareAndSet(this, null, subscription)) {
+                nextSource();
```

Comment:
`nextSource()` is also called in `InnerSubscriber`. And JIT compiler will inline `nextSource()` if it is frequently invoked.

Reply:
If we just call `subscription.request(1)` here, there's no additional round trip to memory because `subscription` is a local variable.
However, if we call `nextSource()`, which calls `upstream.request(1)` in it, there's additional round trip to the memory because `upstream` is a volatile field.
So insn't it better to just call `subscription.request(1)` here? 😉

Reply:
> there's additional round trip to the memory because upstream is a volatile field.

Ah... That makes sense. Agreed.